### PR TITLE
Implement feature 4 : compute balance 

### LINKED ISSUE #4: Issue 4 — Compute balances (equal split) (`Splitter.computeBalances`)
**Goal:** For each expense, split equally among participants and compute net per person.

**Rules:**
- Share = amount / number_of_participants (BigDecimal)
- Each participant owes their share (negative in their net).
- Payer’s net increases by the full amount (common model).
- Sum of all nets ~ 0 (within cents).

**Acceptance Criteria:**
- Works for the provided 3-row sample.
- Handles single-participant case.
- Nets retrievable via `Balance.getNames()` / `getAmount(name)`.

### PR DESCRIPTION
**Pull request by MelonChicken**
### implemented feature 4 : compute balance  #4 
Consider one happy case and two edge case, which resulted build success through the command,`mvn surefire-report:report`

### Implementation Explanation
1. For every expense, the amount is evenly divided among all participants using precise BigDecimal math.
  - The payer’s balance increases by the full amount they paid.
  - Each participant’s balance decreases by their equal share.
  
2. All balances are stored and updated in a shared Balance object using `put()` and `getAmount()` methods.
3. A correction step in each expense distribution ensures the total sum of each net equals `zero`.
4. To ensure that total sum of all net, we double-check the total sum in the end by giving a random participant any rounding remainder.

#### Edge Case
1. Negative amount input case
2. Single participant input case

closes #4

### CODE REVIEW

**Comment by hayat711** [comment]
**Group2** will review this PR [code implemented by group3 for feature #4 ]

File: src/main/java/itm/oss/splitter/Splitter.java
Comment (Gyeongyoon) [inline src/main/java/itm/oss/splitter/Splitter.java:1]: It seems that some unused code has been left commented out.(line 66, 72, 89) If it’s no longer needed, consider removing it to improve readability and keep the codebase clean. 😊

File: src/main/java/itm/oss/splitter/Splitter.java [hunk 1]
```diff
@@ -1,12 +1,91 @@
...
+      }
+      if(total.compareTo(BigDecimal.valueOf(0.00)) != 0) {
+          // decide who will receive the leftover
+          Random random = new Random();
+          String ranName = names.get(random.nextInt(names.size()));
+          result.put(ranName, result.getAmount(ranName).subtract(total));
+
+          return result;
+      } else {
+          return result;
```
Comment (Gyeongyoon) [inline src/main/java/itm/oss/splitter/Splitter.java]: Since both paths return the same value, maybe you can just move the return result; outside the if to make it a bit cleaner.

File: src/main/java/itm/oss/splitter/Splitter.java [hunk 2]
```diff
@@ -1,12 +1,91 @@
...
+          // System.out.println("[after distributing] current total net is "+total);
+      }
+      //Sum nets so total = 0; scale(2) rounding HALF_EVEN when needed.
+
+      for(String person : names) {
+          total = total.add(result.getAmount(person));
+      }
+      if(total.compareTo(BigDecimal.valueOf(0.00)) != 0) {
+          // decide who will receive the leftover
+          Random random = new Random();
```
Comment (ichbinju0) [inline src/main/java/itm/oss/splitter/Splitter.java]: The current implementation uses Random to assign the final rounding error to a random participant. However, SplitterTest expects fixed values in its assertions. So, the test will randomly fail if the rounding error occurs and Random happens to assign it to a participant different from the one the test case expects. I suggest you to replace the random logic. I suggest you to get the list of all participants from the Balance object. → Sort the list → Make a rule for who to assign the rounding error. This will make the method’s output more consistent, and guarantee the test results.

File: src/main/java/itm/oss/splitter/Splitter.java [hunk 3]
```diff
@@ -1,12 +1,91 @@
...
 import java.util.ArrayList;
+import java.util.Random;
 
 public class Splitter {
 
   public Balance computeBalances(ArrayList<Expense> xs) {
-    // TODO (Issue 4): equal split math.
-    // Sum nets so total = 0; scale(2) rounding HALF_EVEN when needed.
-    throw new UnsupportedOperationException("computeBalances() not implemented yet");
+      /**
```
Comment (ichbinju0) [inline src/main/java/itm/oss/splitter/Splitter.java]: I suggest moving this comment block above the method signature to convert it into proper JavaDoc, as it serves as the method's official description. Also, since the feature is now implemented, I recommend deleting the //TODO (Issue 4: ) comment and debugging System.out.println statements. They are no longer needed and harm code readability.

**Review by ichbinju0 — COMMENTED** [review]
Your code is very sophisticated and shows that you've tried to account for every possible scenario. It would be even better with some slight refinement in readability and logical flow.

File: src/test/java/itm/oss/splitter/SplitterTest.java [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+        // net(Alice) = 60 - (60 + 120 + 90) = -210
+        // net(Bob) = 0 + 120 = 120
+        // net(Cara) = 0 + 90 = 90      
+        assertEquals(resultEdgeSingle.getAmount("Alice").setScale(2, RoundingMode.HALF_EVEN),
+                BigDecimal.valueOf(-210.00).setScale(2, RoundingMode.HALF_EVEN));
+        assertEquals(resultEdgeSingle.getAmount("Bob").setScale(2, RoundingMode.HALF_EVEN),
+                BigDecimal.valueOf(120.00).setScale(2, RoundingMode.HALF_EVEN));
+        assertEquals(resultEdgeSingle.getAmount("Cara").setScale(2, RoundingMode.HALF_EVEN),
+                BigDecimal.valueOf(90.00).setScale(2, RoundingMode.HALF_EVEN));
+    }
```
Comment (so-yeon-333) [inline src/test/java/itm/oss/splitter/SplitterTest.java]: I noticed that your team's test file have three test scenarios in the just one big function testComputeBalance().

This is not the best way for unit testing. This makes debugging harder than making three test cases with three different functions.

I would recommend you to split the big function into three small test functions using three @test. Each function should check only one specific thing in the unit testing.

suggestions
```
@Test
@DisplayName("1. Happy-path: All participants share expense equally") void testHappyPathSplit() { }

@Test
@DisplayName("2. Edge Case: Negative expense amount is ignored") void testEdgeCaseNegativeAmount() { }

@Test
@DisplayName("3. Edge Case: Expense with only one participant") void testEdgeCaseSingleParticipant() { }

```

Comment (so-yeon-333) [inline src/test/java/itm/oss/splitter/SplitterTest.java]: I see you are trying to test the 'Single Participant' case, but the setup is confusing.

You modified the names list, and then all expenses use this modified list for their participants. This means Alice is the only participant for every expense. And I'm confused why Bob and Cara pay for something that only Alice shares. What is the specific goal of this edge case? It's hard to understand the test case's situation. What are you wanting to test in this test case? Maybe you could explain the situation more.

**Review by so-yeon-333 — CHANGES_REQUESTED** [review]
Hello, I am so-yeon.

Your team did great, however I found two things in the test file that might be better if you change them.

1.  Unit tests: Split one big function into three @Test methods might be better for maintainabillity and readabillity.
2.  Edge case situation: The case is confusing. Maybe you could explain more about this for readability.

I am requesting changes. Let me know if you see my comments.

File: src/test/java/itm/oss/splitter/SplitterTest.java (hunk 4 above)
Comment (MelonChicken) [reply src/test/java/itm/oss/splitter/SplitterTest.java]: First of all, thanks for the comment to point out the explanation of this test situation. :sunglasses:

In this test case, the intention is to verify the behavior when **each expense has only one participant**.

**In our domain model:**

- participants : people who **share/benefit from the expense**

- payer : the person who **actually paid the money**

Here we assume a scenario where Alice is the only person who actually participates in all expenses, but sometimes Bob or Cara pays on **her behalf**. In other words, Bob and Cara are effectively “lending” money to Alice. I believe that the point of this test case is not about why Alice has to pay all expenses but about the correctness of calculation in the one participant case.

> One of possible situations can be the Alice is the host of the house party so she has to pay all stuffs. Bob and Cara just help Alice's work to buy some stuffs or pay the transportation fee on **her behalf**.

**The goal of this edge case is:**

- to check that when there is **only one participant** for each expense,

the balances are computed correctly so that Alice ends up owing the full amounts that Bob and Cara paid for her.

That’s why the final balances are:

Alice: -210 (owes 120 to Bob and 90 to Cara)

Bob: +120

Cara: +90

I agree that this intention is not obvious from the code alone, so I’ll improve the test name and add a short comment to describe this “single participant / Alice-only” scenario more clearly. Check 928acf8f41dca88ac1e473f632319559ae2f5f32

```java
    @Test
    @DisplayName("3. Edge Case: All expenses have a single participant (Alice)")
    void testEdgeCaseSingleParticipant() {
        // Scenario:
        // - Only Alice is listed as participant for every expense.
        // - Sometimes Bob or Cara pays on behalf of Alice.
        // Expected:
        // - Alice owes the full amounts paid by Bob and Cara.
        //   => Alice: -210, Bob: +120, Cara: +90
...
    }
```

Comment (MelonChicken) [reply src/test/java/itm/oss/splitter/SplitterTest.java]: Thanks for your suggestion to refactor the structure test code!

As you pointed out, the original code was mainly focused on verifying the code result of calculations so the readability of this test was quite poor. :weary:

Since your proposal was already Looks Good To Me, I split the original test into three separate tests based on your suggestion and updated the code as follows. check 928acf8f41dca88ac1e473f632319559ae2f5f32

```suggestion

    @Test
    @DisplayName("1. Happy-path: All participants share expense equally")
    void testHappyPathSplit() {
        ArrayList<Expense> expensesHappyCase = new ArrayList<>();

        ArrayList<String> names = new ArrayList<>();
        names.add("Alice");
        names.add("Bob");
        names.add("Cara");

        Expense expense01 = new Expense("2025-10-01",
                "Alice", BigDecimal.valueOf(70.00),
                "USD",names,"Food","Pizza night");


        Expense expense02 = new Expense("2025-10-02",
                "Bob", BigDecimal.valueOf(120.00),
                "USD",names,"Transport","Taxi + subway");

        Expense expense03 = new Expense("2025-10-03",
                "Cara", BigDecimal.valueOf(90.00),
                "USD",names,"Groceries","Market");

        expensesHappyCase.add(expense01);
        expensesHappyCase.add(expense02);
        expensesHappyCase.add(expense03);

        Balance resultHappyCase = splitter.computeBalances(expensesHappyCase);
        System.out.println("\n[Test for Happy Case : Each participant's net]\n");
        for(String name : names) {
            System.out.println(name + " : " + resultHappyCase.getAmount(name));
        }
        assertEquals(resultHappyCase.getAmount("Alice"), BigDecimal.valueOf(-23.34));
        assertEquals(resultHappyCase.getAmount("Bob"), BigDecimal.valueOf(26.67));
        assertEquals(resultHappyCase.getAmount("Cara"), BigDecimal.valueOf(-3.33));
    }

    @Test
    @DisplayName("2. Edge Case: Negative expense amount is ignored")
    void testEdgeCaseNegativeAmount() {

        ArrayList<String> names = new ArrayList<>();
        names.add("Alice");
        names.add("Bob");
        names.add("Cara");

        ArrayList<Expense> edgeCaseNegative = new ArrayList<>();

        Expense expense04 = new Expense("2025-10-01",
                "Alice", BigDecimal.valueOf(-100.00),
                "USD",names,"Food","Pizza night");


        Expense expense05 = new Expense("2025-10-02",
                "Bob", BigDecimal.valueOf(80.00),
                "USD",names,"Transport","Taxi + subway");

        Expense expense06 = new Expense("2025-10-03",
                "Cara", BigDecimal.valueOf(90.00),
                "USD",names,"Groceries","Market");

        edgeCaseNegative.add(expense04);
        edgeCaseNegative.add(expense05);
        edgeCaseNegative.add(expense06);

        Balance resultEdgeNegative = splitter.computeBalances(edgeCaseNegative);
        System.out.println("\n[Test for Edge Case : negative amount - Each participant's net]\n");
        for(String name : resultEdgeNegative.getNames()) {
            System.out.println(name + " : " + resultEdgeNegative.getAmount(name));
        }
        assertEquals(resultEdgeNegative.getAmount("Alice"), BigDecimal.valueOf(-56.67));
        assertEquals(resultEdgeNegative.getAmount("Bob"), BigDecimal.valueOf(23.34));
        assertEquals(resultEdgeNegative.getAmount("Cara"), BigDecimal.valueOf(33.33));

    }

    @Test
    @DisplayName("3. Edge Case: All expenses have a single participant (Alice)")
    void testEdgeCaseSingleParticipant() {
        // Scenario:
        // - Only Alice is listed as participant for every expense.
        // - Sometimes Bob or Cara pays on behalf of Alice.
        // Expected:
        // - Alice owes the full amounts paid by Bob and Cara.
        //   => Alice: -210, Bob: +120, Cara: +90

        ArrayList<String> names = new ArrayList<>();

        ArrayList<Expense> EdgeCaseSingle = new ArrayList<Expense>();

        names.add("Alice");

        Expense expense07 = new Expense("2025-10-01",
                "Alice", BigDecimal.valueOf(60.00),
                "USD",names,"Food","Pizza night");

        Expense expense08 = new Expense("2025-10-02",
                "Bob", BigDecimal.valueOf(120.00),
                "USD",names,"Transport","Taxi + subway");

        Expense expense09 = new Expense("2025-10-03",
                "Cara", BigDecimal.valueOf(90.00),
                "USD",names,"Groceries","Market");
        EdgeCaseSingle.add(expense07);
        EdgeCaseSingle.add(expense08);
        EdgeCaseSingle.add(expense09);
        Balance resultEdgeSingle = splitter.computeBalances(EdgeCaseSingle);
        System.out.println("\n[Test for Edge Case : Single participants - Each participant's net]\n");

        for(String name : resultEdgeSingle.getNames()) {
            System.out.println(name + " : " + resultEdgeSingle.getAmount(name));
        }
        assertEquals(resultEdgeSingle.getAmount("Alice").setScale(2, RoundingMode.HALF_EVEN), BigDecimal.valueOf(-210.00).setScale(2, RoundingMode.HALF_EVEN));
        assertEquals(resultEdgeSingle.getAmount("Bob").setScale(2, RoundingMode.HALF_EVEN), BigDecimal.valueOf(120.00).setScale(2, RoundingMode.HALF_EVEN));
        assertEquals(resultEdgeSingle.getAmount("Cara").setScale(2, RoundingMode.HALF_EVEN), BigDecimal.valueOf(90.00).setScale(2, RoundingMode.HALF_EVEN));
    }
```

File: src/main/java/itm/oss/splitter/Splitter.java
Comment (GeonKim0422) [reply src/main/java/itm/oss/splitter/Splitter.java:1]: Thank's for your kind review. I've checked unnecessasry lines you remarked and deleted them.

File: src/main/java/itm/oss/splitter/Splitter.java (hunk 1 above)
Comment (GeonKim0422) [reply src/main/java/itm/oss/splitter/Splitter.java]: I have checked the if-else statement and refined as you said.
image

File: src/test/java/itm/oss/splitter/SplitterTest.java (hunk 4 above)
Comment (so-yeon-333) [reply src/test/java/itm/oss/splitter/SplitterTest.java]: Making it to three separate tests looks better than before!! Great Job!! LGTM

Comment (so-yeon-333) [reply src/test/java/itm/oss/splitter/SplitterTest.java]: Thanks a lot for your kind response. Now I get the situation.  Improving the comments look better in readability than before. LGTM!!

**Review by so-yeon-333 — APPROVED** [review]
Looks good to me!  The changed code is easier to read than before. You did a great job!! Thank you for your hard work

**Review by Gyeongyoon — APPROVED** [review]
Thanks for addressing the previous feedback! The changes look great and clearly improve the code — LGTM 👍

File: src/main/java/itm/oss/splitter/Splitter.java (hunk 3 above)
Comment (bigmansinwoo-coder) [reply src/main/java/itm/oss/splitter/Splitter.java]: Thank you for the suggestion!
I moved the explanatory comment block above the method signature and converted it into proper JavaDoc format.
Also, I removed the outdated // TODO (Issue 4) comment and unnecessary System.out.println statements to improve code clarity and readability.

File: src/main/java/itm/oss/splitter/Splitter.java (hunk 2 above)
Comment (bigmansinwoo-coder) [reply src/main/java/itm/oss/splitter/Splitter.java]: I replaced the random assignment with a deterministic rule.
After all balances are computed, I now sort the participant list alphabetically and assign any rounding remainder to the first name in that list.
This ensures consistent and repeatable results for the SplitterTest assertions.

Comment (ichbinju0) [reply src/main/java/itm/oss/splitter/Splitter.java]: Thank you for accepting my comment. I think the code has become more stable. LGTM

File: src/main/java/itm/oss/splitter/Splitter.java (hunk 3 above)
Comment (ichbinju0) [reply src/main/java/itm/oss/splitter/Splitter.java]: Thank you for implementing my comment! Your code is more good to read now. LGTM

**Review by ichbinju0 — APPROVED** [review]
LGTM
